File: native/src/host_resolver.c
/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0.
 */

#include "crt.h"
#include "exports.h"

#include <aws/io/host_resolver.h>

AWS_DOTNET_API
struct aws_host_resolver *aws_dotnet_host_resolver_new_default(struct aws_event_loop_group *elg, int max_hosts) {
    if (!elg) {
        aws_dotnet_throw_exception(AWS_ERROR_INVALID_ARGUMENT, "Invalid EventLoopGroup");
        return NULL;
    }
    struct aws_allocator *allocator = aws_dotnet_get_allocator();
    struct aws_host_resolver *resolver = aws_host_resolver_new_default(allocator, max_hosts, elg, NULL);
    if (resolver == NULL) {
        aws_dotnet_throw_exception(aws_last_error(), "Unable to initialize default host resolver");
    }

    return resolver;
}

AWS_DOTNET_API
void aws_dotnet_host_resolver_destroy(struct aws_host_resolver *resolver) {
    aws_host_resolver_release(resolver);
}

File: aws-crt/IO/HostResolver.cs
/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0.
 */

using System;
using System.Security;
using System.Runtime.InteropServices;

namespace Aws.Crt.IO
{
    public abstract class HostResolver
    {
        [SecuritySafeCritical]
        internal static class API
        {
            public delegate Handle aws_dotnet_host_resolver_new_default(IntPtr eventLoopGroup, int maxHosts);
            public delegate void aws_dotnet_host_resolver_destroy(IntPtr hostResolver);

            public static aws_dotnet_host_resolver_new_default make_new_default = NativeAPI.Bind<aws_dotnet_host_resolver_new_default>();
            public static aws_dotnet_host_resolver_destroy destroy = NativeAPI.Bind<aws_dotnet_host_resolver_destroy>();
        }

        public class Handle : CRT.Handle
        {
            protected override bool ReleaseHandle()
            {
                API.destroy(handle);
                return true;
            }
        }

        public Handle NativeHandle { get; private set; }

        internal HostResolver(Handle handle) 
        {
            this.NativeHandle = handle;
        }
    }

    public sealed class DefaultHostResolver : HostResolver
    {
        public DefaultHostResolver(EventLoopGroup eventLoopGroup, int maxHosts=64)
            : base(API.make_new_default(eventLoopGroup.NativeHandle.DangerousGetHandle(), maxHosts))
        {
        }
    }
}